src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs

```diff
@@ -0,0 +1,488 @@
+                        // We perform the first fast pass only if there are no delegate creations/lambda methods.
+                        // This is due to the fact that tracking which local/parameter points to which delegate creation target
+                        // at any given program point needs needs flow analysis (second pass).
+                        if (!_hasDelegateCreationOrAnonymousFunction)
+                        {
+                            var resultFromOperationBlockAnalysis = SymbolUsageAnalysis.Run(operationBlock, context.OwningSymbol, context.CancellationToken);
+                            if (!resultFromOperationBlockAnalysis.HasUnreadSymbolWrites())
+                            {
+                                // Assert that even slow pass (dataflow analysis) would have yielded no unused symbol writes.
+                                Debug.Assert(!SymbolUsageAnalysis.Run(context.GetControlFlowGraph(operationBlock), context.OwningSymbol, context.CancellationToken)
+                                             .HasUnreadSymbolWrites());
+
+                                hasBlockWithAllUsedSymbolWrites = true;
+                                continue;
+                            }
+                        }
+
+                        // Now perform the slower, precise, CFG based dataflow analysis to identify the actual unused symbol writes.
+                        var cfg = context.GetControlFlowGraph(operationBlock);
+                        var symbolUsageResult = SymbolUsageAnalysis.Run(cfg, context.OwningSymbol, context.CancellationToken);
+                        symbolUsageResultsBuilder.Add(symbolUsageResult);
+
+                        foreach (var (symbol, unreadWriteOperation) in symbolUsageResult.GetUnreadSymbolWrites())
+                        {
+                            if (unreadWriteOperation == null)
+                            {
+                                // Null operation is used for initial write for the parameter from method declaration.
+                                // So, the initial value of the parameter is never read in this operation block.
+                                // However, we do not report this as an unused parameter here as a different operation block
+                                // might be reading the initial parameter value.
+                                // For example, a constructor with both a constructor initializer and body will have two different operation blocks
+                                // and a parameter must be unused across both these blocks to be marked unused.
+
+                                // However, we do report unused parameters for local function here.
+                                // Local function parameters are completely scoped to this operation block, and should be reported per-operation block.
+                                var unusedParameter = (IParameterSymbol)symbol;
```

Comment:
it feels a little risky dependin gon this being an IParameterSymbol

Reply:
Improved the doc comment on `GetUnreadSymbolWrites` to explicitly mention this for clarity.

---
In reply to: [234867795](https://github.com/dotnet/roslyn/pull/30777#discussion_r234867795) [](ancestors = 234867795)

Reply:
thanks!